draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1216 @@
+if retried will surely fail if they fail at those sections.
+
+.. code-block:: python
+
+  from celery import task, poisonous
+  from data_validation_lib import validate_data
+
+  def _calculate(a, b):
+    # Do something
+
+  @task(autoretry_for=(ValueError,))
+  def complex_calculation(a, b):
+    with poisonous():
+      validate_data(a, b)
+
+    # May temporarily raise a ValueError due to some externally fetched
+    # data which is currently incorrect but will be updated later.
+    _calculate()
+
+Not all operations are equal. Some may be retried more than others.
+Some may need to be retried less often.
+
+Currently there are multiple ways to achieve this:
+
+You can separate them to different tasks with a different retry policy:
+
+.. code-block:: python
+
+  from celery import task
+
+  @task(retry_policy={
+    'max_retries': 3,
+    'interval_start': 0,
+    'interval_step': 0.2,
+    'interval_max': 0.2
+  })
+  def foo():
+    second_operation()
+
+  @task(retry_policy={
+    'max_retries': 10,
+    'interval_start': 0,
+    'interval_step': 5,
+    'interval_max': 120
+  })
+  def bar():
+    first_operation()
+    foo.delay()
+
+Or you can wrap each code section in a try..except clause and call
+:py:meth:`celery.app.task.Task.retry`.
+
+.. code-block:: python
+
+  @task(bind=True)
+  def foo(self):
+    try:
+      # first operation
+    except Exception:
+      self.retry(retry_policy={
+        'max_retries': 10,
+        'interval_start': 0,
+        'interval_step': 5,
+        'interval_max': 120
+      })
+
+    try:
+      first_operation()
+    except Exception:
+      self.retry(retry_policy={
+        'max_retries': 10,
+        'interval_start': 0,
+        'interval_step': 5,
+        'interval_max': 120
+      })
+
+    try:
+      second_operation()
+    except Exception:
+      self.retry(retry_policy={
+        'max_retries': 3,
+        'interval_start': 0,
+        'interval_step': 0.2,
+        'interval_max': 0.2
+      })
+
+Those solutions are unnecessarily verbose. Instead, we could use a with clause
+if all we want to do is retry.
+
+.. code-block:: python
+
+  @task
+  def foo():
+    with retry(max_retries=10, interval_start=0, interval_step=5, interval_max=120):
+      first_operation()
+
+    with retry(max_retries=10, interval_start=0, interval_step=5, interval_max=120):
+      second_operation()
+
+By default messages which cannot be re-published will be stored
+in the :ref:`draft/celery-5-high-level-architecture:messages backlog`.
+
+Implementers may provide other fallbacks such as executing the retried task
+in the same worker or abandoning the task entirely.
+
+Health Checks
+~~~~~~~~~~~~~
+
+Health Checks are used in Celery to verify that a worker is able to
+successfully execute a :ref:`task <draft/celery-5-high-level-architecture:Tasks>`
+or a :ref:`service <draft/celery-5-high-level-architecture:Services>`.
+
+The :ref:`draft/celery-5-high-level-architecture:Scheduler` is responsible
```

Comment:
Is this a liveness check/ping command? I think it could be the other way round too, each worker should send heartbeats to a centralized service or publish them to a queue.